core/client/src/client.rs

```diff
@@ -330,6 +330,11 @@ impl<B, E, Block> Client<B, E, Block> where
 		block_builder::BlockBuilder::at_block(parent, &self)
 	}
 
+	/// Get the best block id.
+	pub fn best_block_id(&self) -> error::Result<BlockId<Block>> {
+		Ok(BlockId::Number(self.info()?.chain.best_number))
```

Comment:
this is very vulnerable to reorgs

Reply:
how is it more vulnerable than hash? During reorg the best hash is more likely to change than best number.

Reply:
but it won't refer to the same block anymore. With a hash, you actually have a stable block that you're referencing.

During a reorg the best number always changes too, at least with the longest chain rule.

Reply:
A user of this API will expect that the ID returned by this function will always reference the same block. 

```rust
let best_id = client.best_block_id(&self)?;
let hdr_a = client.header(best_id)?;
// some time later...
let hdr_b = client.header(best_id)?;

assert_eq!(hdr_a, hdr_b);
``` 

The idea that this snippet can fail is frightening.

Reply:
> A user of this API will expect that the ID returned by this function will always reference the same block.

`BlockId` is not designed to always reference the same block as it can be a mere block number.

```rust
let hdr_a = client.header(id)?;
// some time later...
let hdr_b = client.header(id)?;
assert_eq!(hdr_a, hdr_b);
```
can fail fail regardless of where you get `id` from.

As for the second example, that race would still be there if you had atomic `best_block_header`

```rust
let best_id = client.best_block_id(&self)?;

let hdr = client.header(best_id)?;
// import happens here and hdr is not best anymore
```

> // hdr could be a block that was never the best at any point in time.

That's not true. `client.header()` by number only returns best blocks.

Reply:
```
A -> B
```

B is best.

```
  -> B' 
 /
A -> B
```

B still best.

```
   -> B' -> C
  /
A 
  \ 
   -> B
```

C is best, but B' was never best.

Reply:
> BlockId is not designed to always reference the same block as it can be a mere block number.

Yes, this is why it's discouraged to use numbers in general in our codebase. Why should this API encourage it? It's so easily error-prone.

Reply:
Ok, The client should not return `BlockId` in the first place. `BlockId` is for the user of the client API to create as only they can decide if they want to use block numbers or hashes. Will remove this.